src/main/java/net/openhft/chronicle/threads/MediumEventLoop.java

```diff
@@ -222,10 +256,13 @@ public long loopStartNS() {
     public void run() {
         try {
             try (AffinityLock lock = AffinityLock.acquireLock(binding)) {
-                thread = Thread.currentThread();
-                if (thread == null)
-                    throw new NullPointerException();
-                loopStartedAllHandlers();
+                // Make sure nobody's adding a handler while we do this
+                synchronized (addHandlerMutex) {
+                    thread = Thread.currentThread();
+                    if (thread == null)
+                        throw new NullPointerException();
+                    loopStartedAllHandlers();
+                }
```

Comment:
We take the `addHandlerMutex` to make the setting of `thread` and the calling of `loopStarted` on all the handlers atomic.

If an external thread attempts to add a handler while this is going on, it either occurs before all of this (`addHandlerBeforeStart` succeds) or after all of this (we fall back to `addHandlerAfterStart` behaviour)